engine/engine.go

```diff
@@ -524,7 +525,8 @@ forLoop:
 			if mapStatus[status] {
 				if status == resolution.StateWaiting && recheckWaiting {
 					for name, s := range res.Steps {
-						if s.State == step.StateWaiting {
+						// Steps using the batch plugin shouldn't be run again when WAITING
+						if s.State == step.StateWaiting && s.Action.Type != pluginbatch.Plugin.PluginName() {
```

Comment:
I'm not sure why steps are run once again when they reach the WAITING state, but it creates a race condition with the batch plugin. If the last task of the a sub-batch tries to resume its parent, but said parent is being run a second time by the engine, preventing the child task from resuming it.

Reply:
Add this explanation in comment to understand this quickfix/hack